### PR TITLE
chore: update to webpack v5

### LINKED ISSUE #19555: Investigate Webpack 5 Impact
### What would you like?

Investigate upgrading Cypress to webpack v5+.

### Details
The following modules could / will be impacted:

- cypress-webpack-preprocessor
    - peer/devDep v4 & v5
    - MVB to only support webpack v5
- cypress-webpack-batteries-included-preprocessor
    - dep v4 - MVB
- cypress-webpack-dev-server
    - devDep v4

The packages within the Cypress Binary that could / will be impacted:
- webpack-gui
- driver
- extensions
- reporter
- runner
- runner-ct
- server
- ui-component
- web-config

### Why is this needed?

1. Be on the latest webpack version
2. Better Node 17 handling - see #18914 & #19320

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes #19555

### What is in this PR
This PR updates the `cypress` monorepo to use `webpack@5` from `webpack@4` doing the bare minimum updating to get webpack 5 working. This handles the [Test webpack 5 compatibility](https://webpack.js.org/migrate/5/#test-webpack-5-compatibility) and [Upgrade webpack to 5](https://webpack.js.org/migrate/5/#upgrade-webpack-to-5) steps in the migration guide

The main goal of updating to webpack 5 is to get with the recent dependency updates, as well as unblock the electron v25 [upgrade](https://github.com/cypress-io/cypress/pull/26416). Since newer versions of Electron run on Node 18, it causes issues with `--openssl-legacy-provider`, which is fixed by this upgrade (see code removed in b973ca57ad ). See 
* https://github.com/cypress-io/cypress/pull/19094 
* https://github.com/cypress-io/cypress/issues/18914
* https://github.com/cypress-io/cypress/issues/19320

Most the changes needed are well documented inside the code as to why we need them. This includes 

* 2 dependency patches, one to `whatwg-url` and another to `enhanced-resolve` due to polyfilled differences (see code). In addition to these patches and to make sure they are applied correctly, dev dependency patches are not carried over to the build directory (see 53e304fb77).
* [polyfilling](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#automatic-nodejs-polyfills-removed) node built ins for the `webpack.base.config.ts` inside `web-config` for Cypress internally, as well as polyfilling correctly for the `webpack-batteries-included-preprocessor` through [resolve.fallback](https://webpack.js.org/configuration/resolve/#resolvefallback), [ProvidePlugin](https://webpack.js.org/plugins/provide-plugin/), and [DefinePlugin](https://webpack.js.org/plugins/define-plugin/)
* addressing some subtle differences in the `path` polyfill code (see f01f645ebf)
* addressing small issue where first bundled returned promise reference was changing under the hood, making the `webpack-preprocessor` time out (see 1fa6183855). This promise reference seems to be preserved correctly in `webpack@4`, which seems odd since it should not be as the `webpack@5` handles this correctly.
* a few snapshot changes in `system-tests`, mainly due to difference source-map paths and console outputs

##### Performance considerations
@mjhenkes and I looked into the telemetry data on the initial green build of the run. We both feel confident the performance currently is about the same and no major regressions are present.

### What is NOT in this PR
Loader/Plugin updates are not in this PR as they are not needed to update to `webpack@5`. However, we should update to the latest loaders to gain benefits of recent versions, which will be done in a follow up PR

Additionally, there are still build warnings that need to be fixed from the `webpack@5` update. These will also be addressed in a follow up PR to prevent any breakages when deprecated behavior is removed.

Optimizations to the build also need to be performed, which will be addressed in a separate PR. see [clean up configuration](https://webpack.js.org/migrate/5/#clean-up-configuration) in the migration guide. 

### Is this a breaking change to `cypress`?

This is **NOT** a breaking change to `cypress`, however it does change the way cypress is bundled under the hood. This should not break end users. However, since [polyfills have been removed](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#automatic-nodejs-polyfills-removed) in `webpack@5`, we need to make sure that dependencies that are bundled via the [ProvidePlugin](https://webpack.js.org/plugins/provide-plugin/) are available dependencies in a user's project if `cypress` is installed. Otherwise, bundling _could_ fail. This means that `buffer` and `process` need to be installed with the CLI in the case a user does NOT have `buffer` or `process` as a fallback in their dependency tree. Other polyfills that are set through [resolve.fallback](https://webpack.js.org/configuration/resolve/#resolvefallback) are installed locally to the dependency (such as `webpack-batteries-included-preprocessor`) and can be discovered at processing time without having to hoist/install dependencies in the users root `node_modules` (see 697f8bf8a3).

The other aspect that might be considered breaking is the subtle changes to the node built ins from `webpack@4` to `webpack@5`. We know some of the APIs have changed slightly to the built ins. We have discussed this at length at Cypress, and will NOT be considering this a breaking change. The likelihood users are broken by built in changes is incredibly small, and in the case this is a problem, the user can either:
* update their test code to use the correct built ins
* process their spec code with [@cypress/webpack-preprocessor](https://www.npmjs.com/package/@cypress/webpack-preprocessor) with `webpack@4`

If users are using the `file:preprocessor` hook inside `plugins`, users can still use `webpack@4` still and will not be broken, even though cypress is bundled with webpack 5. The only stipulation to continuing using `webpack@4` is that users remain on Node 16 as `webpack@4` has this [issue that will not be backpatched](https://github.com/webpack/webpack/issues/14532#issuecomment-947513562) from webpack. We recommend users upgrade to recent versions of webpack and node, as Node 16 will be [EOL in Q3 of 2023](https://nodejs.org/en/blog/announcements/nodejs16-eol)

#### Do we need to publish new major versions of `webpack-preprocessor`, `webpack-dev-server`, and `webpack-batteries-included-preprocessor`? 

The only package that will require a major version publish from this change will be `@cypress/webpack-batteries-included-preprocessor` as it will be shipping with `webpack@5` from now on. The breaking changes will be included in this PR, but will **NOT** be published in this PR as it is flagged as a `chore` and NOT `BREAKING_CHANGE`. The `BREAKING_CHANGE` commit will  be handled in a subsequent PR (see https://github.com/cypress-io/cypress/pull/27469). The strategy for release of this PR will be:

* merge this PR as a `chore` into `develop`. Nothing is released
* https://github.com/cypress-io/cypress/pull/27469 is merged into `develop` as a `BREAKING_CHANGE` to `@cypress/webpack-batteries-included-preprocessor`. This publishes the new version of the package to npm.

### Steps to test

Give it a run for yourself locally and make sure everything is good on your end, as well as downloading and testing the binary. We have a [green circle build](https://app.circleci.com/pipelines/github/cypress-io/cypress/55028) for CI testing to make sure everything is good there.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by lmiller1990** [comment]
This should also add support for many historic issue that had "fixes", now those will work out of the box. Example issues:

- https://github.com/cypress-io/cypress/issues/22788
- https://github.com/cypress-io/cypress/issues/25300

And any others relating to wepback compilation errors and newer JS syntax.

File: cli/test/lib/tasks/dependency_spec.js
Comment (lmiller1990) [inline cli/test/lib/tasks/dependency_spec.js:6]: I have some questions about this. If webpack 4 polyfilled all core Node.js libs out of the box, does that mean you could (and can) 

- install cypress 12
- write a spec that imports `fs`
- have it run successfully (eg, `fs` won't error out)

I haven't tried this yet, but worth considering. I know we don't do this in our code base, but I'm sure some code bases rely on those polyfills (especially `crypto`).

I wonder if we need to polyfill ALL the core modules to maintain backwards compatibility.

File: cli/test/lib/tasks/dependency_spec.js
Comment (lmiller1990) [inline cli/test/lib/tasks/dependency_spec.js:22]: Should we assert that 

```js
require('node:buffer') !== require('buffer')
```

That would verify the `Buffer` you get is not the core Node.js one. You can use `require('node:<module>')` to get the core one.

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (lmiller1990) [inline npm/webpack-batteries-included-preprocessor/index.js:173]: Oh, maybe we do polyfill them all. My previous comment may be irrelevant.

Big breaking change here - who knows if these implementations match whatever webpack 4 was doing.

Tracking down the right polyfill here must have been complex.

File: npm/webpack-preprocessor/index.ts
Comment (lmiller1990) [inline npm/webpack-preprocessor/index.ts:329]: Nice job figuring this out 💯

File: packages/driver/src/cy/commands/files.ts
Comment (lmiller1990) [inline packages/driver/src/cy/commands/files.ts:113]: Should we take this chance to stop importing `path`  in the driver, and just copy-paste a minimal `basename` function? I don't think driver should be coupled to a polyfill, or webpack for that matter. We bundle a huge polyfill for a tiny amount of behavior.

In fact, we should be able to eliminate webpack entirely from our build process pretty soon. Only driver and reporter use it, but those could go to Vite pretty soon.

Comment (lmiller1990) [reply packages/driver/src/cy/commands/files.ts:113]: We will still need a buffer implementation, but we could explicitly import that, instead of relying on the bundler to do it for us.

File: packages/web-config/webpack.config.base.ts
Comment (lmiller1990) [inline packages/web-config/webpack.config.base.ts]: Curious what we are ignoring? I remember mocha having some weird issues, but forgot what they were.

File: patches/enhanced-resolve+5.15.0.dev.patch
Comment (lmiller1990) [inline patches/enhanced-resolve+5.15.0.dev.patch:10]: Amazing how many things Yarn 3 breaks :|

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (lmiller1990) [inline npm/webpack-batteries-included-preprocessor/index.js:143]: (unrelated to this line of code, general comment) This should give us out of the box support for things like 

- optional chaining `?.`
- class properties:

```js
class Foo {
  bar = '123'
}
```

Maybe we can add a few tests to make sure this is the case?

Related:

- #22788
- #25300

**Review by lmiller1990 — COMMENTED** [review]
Left some comments / questions, looks good!

Unresolved question: how to release this? I think this will be a Cy 13 breaking change. Migration path isn't clear, though - if there's some way to modify the webpack config without installing `@cypress/webpack-preprocessor`, this is going to be hard.

I suspect this IS NOT a problem. The reason is [the docs say](https://github.com/cypress-io/cypress/tree/develop/npm/webpack-preprocessor):


> Note that installing @cypress/webpack-preprocessor is also required. This allows you to update its version separately from this wrapper. `npm install --save-dev @cypress/webpack-batteries-included-preprocessor @cypress/webpack-preprocessor`

So anyone who wanted to this is on the OLD version of preprocessor. Since we will make this breaking, it'll bump the major. If they upgrade, they'll need to make changes. We might need some notes on how to do that, but they could just refer to the webpack 4-5 migration guide.

File: packages/web-config/webpack.config.base.ts
Comment (AtofStryker) [reply packages/web-config/webpack.config.base.ts]: I am not sure here to be honest. I just migrated the config option over to the v5 equivalent.

File: cli/test/lib/tasks/dependency_spec.js
Comment (AtofStryker) [reply cli/test/lib/tasks/dependency_spec.js:22]: that might be a good idea, considering we really want to just make sure the `buffer` is installed in the package through the `package.json` and its accessible

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (AtofStryker) [reply npm/webpack-batteries-included-preprocessor/index.js:173]: > Oh, maybe we do polyfill them all. My previous comment may be irrelevant.

> Big breaking change here - who knows if these implementations match whatever webpack 4 was doing.

This for sure breaks the `webpack-batteries-included-preprocessor` . In this case, we polyfill most of them and are fairly explicit about the ones we don't in the config. We also have a few tests in place since just about everything goes through the preprocessor. The tests that we do check for the polyfills are [this system test in plugins_spec.js](https://github.com/cypress-io/cypress/blob/develop/system-tests/test/plugins_spec.js#L213) which uses [this spec](https://github.com/cypress-io/cypress/blob/develop/system-tests/projects/e2e/cypress/e2e/node_builtins.cy.js) and checks for just about every polyfill in this [gist](https://gist.github.com/ef4/d2cf5672a93cf241fd47c020b9b3066a#list-of-polyfill-packages-that-were-used-in-webpack-4) except for the ones starting with `_`, in particular the `readable-stream` dependencies. There is also [this unit test](https://github.com/cypress-io/cypress/blob/develop/npm/webpack-batteries-included-preprocessor/test/fixtures/node_builtins_spec.js) that pretty much tests the same thing. At least for the built ins we are want to convey we support we should be covered.

> Tracking down the right polyfill here must have been complex.

this [gist](https://gist.github.com/ef4/d2cf5672a93cf241fd47c020b9b3066a#list-of-polyfill-packages-that-were-used-in-webpack-4) for sure helped to figure out what was going on

**Comment by AtofStryker** [comment]
> Unresolved question: how to release this? I think this will be a Cy 13 breaking change. Migration path isn't clear, though - if there's some way to modify the webpack config without installing `@cypress/webpack-preprocessor`, this is going to be hard.

This is the big thing to figure out here. Unless there is a way to ship new major versions of the `npm` dependencies without shipping a major version of the binary, This will need to go in with v13. And if we aren't sure on any pieces, this might be the safest bet.

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (AtofStryker) [reply npm/webpack-batteries-included-preprocessor/index.js:143]: I can take a look into this and see if I can add a few tests

File: packages/driver/src/cy/commands/files.ts
Comment (AtofStryker) [reply packages/driver/src/cy/commands/files.ts:113]: I agree that we should limit down the use of polyfills we don't really need, especially ones that large. Probably in this case we could just import the tree shakable version from the polyfill and exclude it from webpack. However, since we already have a lot going on here, I'd prefer not to touch it in this PR and keep things as 1:1 as possible. After we get something in, we can create a follow up to strip down some of this, which makes the changeset a bit smaller and clearer.

File: cli/test/lib/tasks/dependency_spec.js
Comment (AtofStryker) [reply cli/test/lib/tasks/dependency_spec.js:6]: I think it depends. `fs` is probably a bad example here since we intentionally leave it empty in the `webpack-batteries-included-preprocessor` and don't expect that built in to be available. But for something like `crypto` for your example, you could do all those steps and it will not error out since we install the polyfill. I think you talked about it in another spot in this PR, but [this comment explains](https://github.com/cypress-io/cypress/pull/27438#discussion_r1281259701) it for anyone else reading

File: packages/driver/src/cy/commands/files.ts
Comment (lmiller1990) [reply packages/driver/src/cy/commands/files.ts:113]: Sounds good. I don't know how much of `path` we use but if we only need one function we can just grab the code for it, that would be fine too - let's revisit this after.

**Review by lmiller1990 — CHANGES_REQUESTED** [review]
I am happy with the code, I am going to request changes to BLOCK this until we have a decision on merging / release - this is definitely not `chore` and most likely needs to be retargeted to target the v13 release branch

File: packages/web-config/webpack.config.base.ts
Comment (AtofStryker) [reply packages/web-config/webpack.config.base.ts]: Ah I found it. I don't think its working properly though so I will get this fixed.
Screenshot 2023-08-02 at 7 48 10 AM

Comment (AtofStryker) [reply packages/web-config/webpack.config.base.ts]: config should be 
```
const ignoreWarnings = [{
  module: /node_modules\/mocha\/lib\/mocha.js/,
}]
```

**Comment by lmiller1990** [comment]
I think it's possible to ship this early but bundle the older version in the binary. If Cy 13 is going out this month, it's not worth it, if we expect delays, we should look into it, though.

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (AtofStryker) [reply npm/webpack-batteries-included-preprocessor/index.js:143]: @lmiller1990 I took a look into both of these. The first looks like its an issue with not transpiling `node_modules`? And I couldnt reproduce the later with Webpack 4. Did you have any luck?

Comment (AtofStryker) [reply npm/webpack-batteries-included-preprocessor/index.js:143]: @lmiller1990 I took a look into both of these. The first looks like its an issue with not transpiling `node_modules`? And I couldnt reproduce the later with Webpack 4. Did you have any luck?

File: packages/web-config/webpack.config.base.ts
Comment (AtofStryker) [reply packages/web-config/webpack.config.base.ts]: fixed in 52312cf610

**Comment by AtofStryker** [comment]
@lmiller1990 I updated the PR description to describe how we are going to release this. We will first merge this in as a `chore`, with a follow up PR that will apply the `BREAKING_CHANGE` commit to `@cypress/webpack-batteries-included-preprocessor`. We are not concerned about built ins breaking and in the very rare case it does impact a user, leveraging the npm preprocessors are an option. I will work on getting the breaking change PR ready for @cypress/webpack-batteries-included-preprocessor` and will update you and the description when ready

**Comment by AtofStryker** [comment]
@lmiller1990 linked https://github.com/cypress-io/cypress/pull/27469 in the PR description in draft. I think this is ready for a re review 😄

**Review by chrisbreiding — APPROVED** [review]

File: cli/CHANGELOG.md
Comment (jennifer-shehane) [inline cli/CHANGELOG.md]:
```suggestion
- Upgraded [`webpack`](https://www.npmjs.com/package/webpack) from `v4` to `v5`. This means that we are now bundling your `e2e` tests with webpack 5. We don't anticipate this causing any noticeable changes. However, if you'd like to keep bundling your `e2e` tests with wepback 4 you can use the same process as before by pinning [@cypress/webpack-batteries-included-preprocessor](https://www.npmjs.com/package/@cypress/webpack-batteries-included-preprocessor) to `v2.x.x` and hooking into the [file:preprocessor](https://docs.cypress.io/api/plugins/preprocessors-api#Usage) plugin event. This will restore the previous bundling process. Additionally, if you're using [@cypress/webpack-batteries-included-preprocessor](https://www.npmjs.com/package/@cypress/webpack-batteries-included-preprocessor) already, a new version has been published to support webpack `v5`.
```

**Review by jennifer-shehane — COMMENTED** [review]
@AtofStryker Suggested changes for the changelog wording.

File: cli/CHANGELOG.md
Comment (lmiller1990) [inline cli/CHANGELOG.md:12]: Would they also need to install webpack@4 to keep things working the same?

File: npm/webpack-batteries-included-preprocessor/index.js
Comment (lmiller1990) [reply npm/webpack-batteries-included-preprocessor/index.js:143]: Hm - looks like those issues are not related to webpack 5. Disregard this.

**Review by lmiller1990 — APPROVED** [review]
Looks good to me.

File: cli/CHANGELOG.md
Comment (AtofStryker) [reply cli/CHANGELOG.md:12]: not in this case because `v2` of `webpack-batteries-included-preprocesor` ships with 4. its not backwards compatible unlike the regular `webpack-preprocessor`

**Comment by AtofStryker** [comment]
jobs are passing but for whatever reason the snyk workflows wont run. We are going to admin merge this branch is barring everything else in CI is passing. Passing snyk jobs from previous runs to show no vulnerabilities to build deps https://github.com/cypress-io/cypress/actions/runs/5730373496 & https://github.com/cypress-io/cypress/actions/runs/5730373491

**Pull request merged**
